Fix configuration reloading

coolljt0725 opened the pull request:
There are five options 'debug' 'labels' 'cluster-store' 'cluster-store-opts'
and 'cluster-advertise' that can be reconfigured, configure any of these
options should not affect other options which may have configured in flags.
But this is not true, for example, I start a daemon with -D to enable the
debugging, and after a while, I want reconfigure the 'label', so I add a file
'/etc/docker/daemon.json' with content '"labels":["test"]' and send SIGHUP to daemon
to reconfigure the daemon, it work, but the debugging of the daemon is also diabled.
I don't think this is a expeted behaviour.
This patch also have some minor refactor of reconfiguration of cluster-advertiser.
Enable user to reconfigure cluster-advertiser without cluster-store in config file
since cluster-store could also be already set in flag, and we only want to reconfigure
the cluster-advertiser.

Signed-off-by: Lei Jitang leijitang@huawei.com
ping @calavera 
I'll add tests if we want this.

Fixes #20954

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22107/console):

~~~console

---


Ping statistics for 10.0.0.13:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 0ms, Average = 0ms
ERROR: Failed with exitcode 1 at Tue Feb 23 11:32:14 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson6688279179356007243.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24077/console):

~~~console

---
	github.com/docker/docker/builder/dockerignore	0.003s	coverage: 93.3% of statements
?   	github.com/docker/docker/cli	[no test files]
ok  	github.com/docker/docker/cliconfig	0.005s	coverage: 91.8% of statements
ok  	github.com/docker/docker/container	0.147s	coverage: 10.8% of statements
?   	github.com/docker/docker/contrib/apparmor	[no test files]
?   	github.com/docker/docker/contrib/docker-device-tool	[no test files]
?   	github.com/docker/docker/contrib/httpserver	[no test files]
--- FAIL: TestDaemonReloadLabels (0.00s)
	daemon_test.go:327: Expected daemon label `foo:baz`, got foo:bar
--- FAIL: TestDaemonDiscoveryReload (0.00s)
	daemon_test.go:384: expected [127.0.0.1:5555], got [127.0.0.1:3333]
--- FAIL: TestDaemonDiscoveryReloadFromEmptyDiscovery (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x30 pc=0x508c63]

goroutine 47 [running]:
testing.tRunner.func1(0
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15289/console):

~~~console

---
	github.com/docker/docker/builder/dockerignore	0.003s	coverage: 93.3% of statements
?   	github.com/docker/docker/cli	[no test files]
ok  	github.com/docker/docker/cliconfig	0.011s	coverage: 91.8% of statements
ok  	github.com/docker/docker/container	0.157s	coverage: 10.8% of statements
?   	github.com/docker/docker/contrib/apparmor	[no test files]
?   	github.com/docker/docker/contrib/docker-device-tool	[no test files]
?   	github.com/docker/docker/contrib/httpserver	[no test files]
--- FAIL: TestDaemonReloadLabels (0.00s)
	daemon_test.go:327: Expected daemon label `foo:baz`, got foo:bar
--- FAIL: TestDaemonDiscoveryReload (0.00s)
	daemon_test.go:384: expected [127.0.0.1:5555], got [127.0.0.1:3333]
--- FAIL: TestDaemonDiscoveryReloadFromEmptyDiscovery (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x30 pc=0x508c63]

goroutine 44 [running]:
testing.tRunner.func1(0
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/2210/console):

~~~console

---
056s
?   	github.com/docker/docker/builder/dockerfile/parser/dumper	[no test files]
ok  	github.com/docker/docker/builder/dockerignore	0.044s
?   	github.com/docker/docker/cli	[no test files]
ok  	github.com/docker/docker/cliconfig	0.053s
ok  	github.com/docker/docker/container	0.295s
?   	github.com/docker/docker/contrib/apparmor	[no test files]
?   	github.com/docker/docker/contrib/docker-device-tool	[no test files]
?   	github.com/docker/docker/contrib/httpserver	[no test files]
--- FAIL: TestDaemonReloadLabels (0.00s)
	daemon_test.go:327: Expected daemon label `foo:baz`, got foo:bar
--- FAIL: TestDaemonDiscoveryReload (0.00s)
	daemon_test.go:384: expected [127.0.0.1:5555], got [127.0.0.1:3333]
--- FAIL: TestDaemonDiscoveryReloadFromEmptyDiscovery (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x18]

goroutine 47 [running]:
github_com_docker_docker_daemon.Tes
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6561/console):

~~~console

---
o:490: DockerSuite.TestEventsRename	0.563s
PASS: docker_cli_events_test.go:439: DockerSuite.TestEventsResize	0.609s
PASS: docker_cli_events_unix_test.go:222: DockerSuite.TestEventsStreaming	0.517s
PASS: docker_cli_events_test.go:19: DockerSuite.TestEventsTimestampFormats	2.198s
PASS: docker_cli_events_test.go:500: DockerSuite.TestEventsTop	0.633s
PASS: docker_cli_events_test.go:46: DockerSuite.TestEventsUntag	2.659s

----------------------------------------------------------------------
FAIL: docker_cli_exec_test.go:22: DockerSuite.TestExec

docker_cli_exec_test.go:28:
    c.Assert(out, checker.Equals, "test")
... obtained string = ""
... expected string = "test"


----------------------------------------------------------------------
PASS: docker_cli_exec_test.go:68: DockerSuite.TestExecAfterContainerRestart	0.836s
PASS: docker_api_exec_test.go:50: DockerSuite.TestExecApiCreateContainerPaused	0.367s
PASS: docker_api_exec_test.go:18: DockerSuite.TestExecApiCreateNoCmd	0.323s
PASS: docker_api_exec_test.go:30: DockerSuite
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22111/console):

~~~console

---


Ping statistics for 10.0.0.13:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
ERROR: Failed with exitcode 1 at Tue Feb 23 12:30:49 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson6450099736921863314.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6565/console):

~~~console

---
pkg-config 	python-dev 	python-mock 	python-pip 	python-websocket 	s3cmd=1.5.0* 	ubuntu-zfs 	xfsprogs 	libzfs-dev 	tar 	--no-install-recommends 	&& ln -snf /usr/bin/clang-3.8 /usr/local/bin/clang 	&& ln -snf /usr/bin/clang++-3.8 /usr/local/bin/clang++' returned a non-zero code: 100
Unable to find image 'docker:66aa007-userns' locally
Pulling repository docker.io/library/docker
Tag 66aa007-userns not found in repository docker.io/library/docker
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson5693970814747856579.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15292/console):

~~~console

---
Suite.TestDaemonLoggingDriverDefaultOverride	1.359s
PASS: docker_cli_daemon_test.go:1217: DockerDaemonSuite.TestDaemonLoggingDriverNone	1.323s
/go/src/github.com/docker/docker/hack/make/test-integration-cli: line 18: 11893 Terminated              ( bundle .integration-daemon-start; bundle .integration-daemon-setup; bundle_test_integration_cli; bundle .integration-daemon-stop ) 2>&1
     11894                       | tee -a "$DEST/test.log"
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson3261504283634880693.sh
+ set +x


---

~~~

calavera commented:
Please add test cases and docs so users know how this behaves.

coolljt0725 committed:
Fix configuration reloading

There are five options 'debug' 'labels' 'cluster-store' 'cluster-store-opts'
and 'cluster-advertise' that can be reconfigured, configure any of these
options should not affect other options which may have configured in flags.
But this is not true, for example, I start a daemon with -D to enable the
debugging, and after a while, I want reconfigure the 'label', so I add a file
'/etc/docker/daemon.json' with content '"labels":["test"]' and send SIGHUP to daemon
to reconfigure the daemon, it work, but the debugging of the daemon is also diabled.
I don't think this is a expeted behaviour.
This patch also have some minor refactor of reconfiguration of cluster-advertiser.
Enable user to reconfigure cluster-advertiser without cluster-store in config file
since cluster-store could also be already set in flag, and we only want to reconfigure
the cluster-advertiser.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
@calavera  test added

calavera commented:
Thanks @coolljt0725 

Change LGTM, but it needs a rebase.

calavera commented on daemon/daemon.go:
```diff
@@ -1602,24 +1602,37 @@ func (daemon *Daemon) initDiscovery(config *Config) error {
+	if config.IsValueSet("label") {
+		daemon.configStore.Labels = config.Labels
+	}
+	if config.IsValueSet("debug") {
```
This is not necessary, but I'm okay with adding it for consistency, that field is only in use when the server starts.

coolljt0725 replied:
Yes, it's not a necessary, but for consistency

coolljt0725 commented:
@calavera  rebased

calavera commented:
@runcom, @vdemeester, @MHBauer can you take a look at this? It should be ready to merge :metal:

coolljt0725 commented:
There are some docs change, also ping @thaJeztah for the docs review, my english is not good(still learning :-)), there may have some issues in docs change

runcom commented on daemon/daemon.go:
```diff
@@ -1602,24 +1602,37 @@ func (daemon *Daemon) initDiscovery(config *Config) error {
-	newAdvertise, err := parseClusterAdvertiseSettings(config.ClusterStore, config.ClusterAdvertise)
-	if err != nil && err != errDiscoveryDisabled {
-		return err
+	var err error
```
could you move this down right where it's needed? line 1622

coolljt0725 replied:
@runcom  The `err` is being used in line 1641 `if err == errDiscoveryDisabled `, so it has to been defined   out of the `if` block

runcom commented:
just a nit, LGTM otherwise ping @vdemeester

vdemeester commented:
LGTM 🐱

thaJeztah commented:
Adding for consideration in a patch release (if we have one)
